### ISSUE TITLE
Socket support for TCP_QUICKACK

### ISSUE THREAD

**Issue by felipepessoto** [dotnet/runtime#798]
[AB#1117049](https://devdiv.visualstudio.com/10e66e43-9645-4201-b128-0fdc3769cc17/_workitems/edit/1117049)
`System.Net.Sockets.Socket` currently implements cross platform support for a large number of manually configurable socket options. We allow many TCP specific options, but do not support the standard `TCP_QUICKACK` option that is supported in both Windows and Unix.

## Rationale and Usage
Nagles algorithm and TCP delayed ack are both used to solve similar problems around network congestion. Both are often enabled by default, but the interactions between them can cause extreme and unnecessary latency. For more information see [this blog](`TCP_QUICKACK`) on MSDN, or this [SD comment](https://developers.slashdot.org/comments.pl?cid=14515105&sid=174457) by John Nagle.

By enabling support for `TCP_QUICKACK`, we allow performance oriented developers to choose how to balance network congestion and latency.

The new option would be set via the standard SetSocketOption API. As the option is already supported natively on all supported platforms, the changes required would be very minimal.

## Proposed API
```csharp
namespace System.Net.Sockets
{
    // Defines socket option names for the <see cref='System.Net.Sockets.Socket'/> class.
    public enum SocketOptionName
    {
        #region SocketOptionLevel.Tcp
        TcpQuickAck = 12,
        #endregion
    }
}
```

[[API proposal added above by @rmkerr. Original issue below]] 

# Original Proposal

I'm not sure if Windows support this at socket level as Linux does. But delayed ack does not works well when the remote endpoint is using Nagle Algorithm (TCP_NODELAY = false).

Reference: https://blogs.technet.microsoft.com/nettracer/2013/01/05/tcp-delayed-ack-combined-with-nagle-algorithm-can-badly-impact-communication-performance/

**Comment by JC3** [comment]
Would this work as a hack for now?

    socket.SetSocketOption(SocketOptionLevel.Tcp, (SocketOptionName)12, 1);

**Comment by scalablecory** [comment]
I think we need someone to dive into these options and do some prototyping on Linux and Windows to understand if we can make this portable.

**Comment by github-cygwin** [comment]
Given that socket option 12 is called TCP_CONGESTION_ALGORITHM, taking values from 0..7, wouldn't it make more
sense to assume that TCP_DELAY_FIN_ACK, value 13, a boolean option, is the WinSock equivalent to TCP_QUICKACK?
The name seems to indicate a relationship, but the documentation is inexistent as well :(

**Comment by juliusfriedman** [comment]
> I think we need someone to dive into these options and do some prototyping on Linux and Windows to understand if we can make this portable.

This would be done by either runtime checks of `ifdef` on the rid but I highly suggest you keep it runtime since technically kernel modules can be loaded and unloaded while processes are running...

**Comment by antonfirsov** [comment]
I did carry out an experiment based on *Case Study 1* in [this article](https://support.microsoft.com/en-us/help/214397/design-issues-sending-small-data-segments-over-tcp-with-winsock). On Windows I was able to disable delayed acks and benchmark significant improve in throughput using the `SIO_TCP_SET_ACK_FREQUENCY` option:

```C#
private const int SIO_TCP_SET_ACK_FREQUENCY = unchecked((int)0x98000017);
private static readonly byte[] Dummy = new byte[0];

private static void DisableDelayedAck(Socket socket)
{
    socket.IOControl(SIO_TCP_SET_ACK_FREQUENCY, BitConverter.GetBytes(1), Dummy);
}
```

I tried setting `TCP_QUICKACK` by calling `socket.SetSocketOption(SocketOptionLevel.Tcp, (SocketOptionName)12, 1)` on both Windows and Ubuntu 18.04, but got a `SocketException` with `SocketError.OperationNotSupported` on both platforms.

In case of Linux, this is quite surprising since the kernel 2.4.4+ is expected to support the option, and the value should be 12. Maybe I'm doing something wrong, will keep experimenting.

**Comment by stephentoub** [comment]
> I tried setting TCP_QUICKACK by calling socket.SetSocketOption(SocketOptionLevel.Tcp, (SocketOptionName)12, 1) on both Windows and Ubuntu 18.04, but got a SocketException with SocketError.OperationNotSupported on both platforms. In case of Linux, this is quite surprising since the kernel 2.4.4+ is expected to support the option, and the value should be 12. Maybe I'm doing something wrong, will keep experimenting.

SetSocketOption takes a SocketOptionName, with values that don't necessarily map to the corresponding values on each platform (since each platform often has different values from each other).  As such, the values go through a translation layer in the PAL, so unless you updated that as well, it's going to try to translate "12" into whatever the PAL says that maps to, or fail if there's no mapping.

If you just want to try to pass "12" directly for experimentation purposes, try using SetRawSocketOption instead.

**Comment by antonfirsov** [comment]
I did some investigations, sharing my results:

- `TCP_QUICKACK` is Linux-specific. According to the manual, and [the analysis in this article (point 3)](https://www.programmersought.com/article/7059675609/) the options is not permanent, therefore it needs to be reset after each `recv` call. In my very basic experiment it persisted though.

- In .NET 5 it can be set with `SetRawSocketOption`:

```C#
socket.SetRawSocketOption((int) SocketOptionLevel.Tcp, 12, BitConverter.GetBytes(1));
```
Note: In my experiments it improved throughput on Linux.

- Setting `TCP_QUICKACK == 12` doesn't work on Windows and leads to an error. (@github-cygwin neither does `TCP_DELAY_FIN_ACK == 13` work -- it seems to be something completely unrelated)
- On Windows it's possible to disable delayed ACKs by changing `SIO_TCP_SET_ACK_FREQUENCY`. See my previous https://github.com/dotnet/runtime/issues/798#issuecomment-653698016 for more details. Unlike `TCP_QUICKACK`, this option has to be set only once.
- On macOS, there is a `sysctl` option `net.inet.tcp.delayed_ack` but I wasn't able to find a way to control it for individual sockets.

#### Conclusion:

I think it's really questionable whether we want to provide a cross platform option for disabling delayed acks. The main issue is the difference between the non-persistent nature of `TCP_QUICKACK` on Linux vs. the persistent behavior of `SIO_TCP_SET_ACK_FREQUENCY` on Windows. Since we likely don't want to do an expensive syscall after each receive on Linux just to guarantee consistence for a corner case, the only way is to implement a setting that behaves inconsistently accross platforms.

I wonder if someone with deeper Linux knowledge can confirm or disprove my conclusions on `TCP_QUICKACK`. @tmds @wfurt?

**Comment by github-cygwin** [comment]
>     * Setting `TCP_QUICKACK == 12` doesn't work on Windows and leads to an error. (@github-cygwin neither does `TCP_DELAY_FIN_ACK == 13` work -- it seems to be something completely unrelated)

@antonfirsov, can you pleae outline which error that is?  Setting this option works, afaics, so are you talking about a later error in read/write?  Which error code is returned?  And on which OS are you testing?

Thanks,
Corinna

**Comment by antonfirsov** [comment]
@github-cygwin trying to threat option 12 as `TCP_QUICKACK` resulted in `OperationNotSupported` for me (I know this is unexpected, I may try again and post code, but don't think it will change the big picture, it's a different switch anyways.). Setting values 0 or 1 on `TCP_DELAY_FIN_ACK (13)` does not lead to an error, but does not have an impact on delayed ack behavior.

**Comment by juliusfriedman** [comment]
Whats wrong with my proposal for `CongestionAlgorithmn` et al?

**Comment by antonfirsov** [comment]
@juliusfriedman `TCP_CONGESTION_ALGORITHM (12)` seems to be an undocumented Windows feature.

**Comment by juliusfriedman** [comment]
@antovfirsov 

On some Windows Server versions I believe it controls various attributes of the TCP transport known as congestion control see [Here](https://en.wikipedia.org/wiki/TCP_congestion_control)I believe you should have a look since it controls things such as ack timeout and duplicate ack etc.

Why are we disabling something specific like delayed ack while not having options for retransmission etc al so that people can use their own algorithmic when required for their applications, they can already use interop or set socket options to do so right?

Typing of such we should really have methods for keep alive as depending on what is in transmission under TCP the application layer might require it.(Justlike Delayed Ack, Retransmission etc)

**Comment by antonfirsov** [comment]
> Why are we disabling something specific like delayed ack while not having options for retransmission etc al so that people can use their own algorithmic means when required for their applications.

A few arguments:
- The number of users requesting the former seems to be quite high, while I don't see strong need for the latter. 
- It's not a good idea to expose undocumented Windows features, unless we have a strong proof they actually work, and at least some confirmation from the Windows TCP team. These are not trivial things to implement and maintain. (Although much simpler, it looks like we even won't be able to provide a cross-platform option for quickack.)
- I don't see options to control these for individual sockets on other platforms. The posts you shared are working with global `sysctl` options.

```C#
(System.Net.Sockets.SocketOptionName)(Common.Extensions.OperatingSystemExtensions.IsWindows ? 13 : 12); //Double check this
```
According to my experiments, these options are not equivalent. `TCP_DELAY_FIN_ACK` is undocumented stuff, we have no information nor empiric proof on what it does.

**Comment by juliusfriedman** [comment]
Hence why I recommended what I did, if you give people the toggles to control the Retransmission time they can sort of implement these algorithms on their own.

On platform where  `sysctl` is not available there is definitely another suitable way to accomplish the same thing but how becomes a question for that distro. What are you trying to accomplish with `sysctl`?

That would be because congestion control algorithms toggle a bunch of switches which is why something like `sysctl` is used to perform the change system wide.

On the individual sockets this would be equal to setting the options manually at the socket level...

Thats because Windows uses the TCP ACK Frequency (I believe) to achieve the same thing: [See](https://support.microsoft.com/en-ph/help/328890/new-registry-entry-for-controlling-the-tcp-acknowledgment-ack-behavior)

Hence again why I recommended giving users all the knobs.

**Comment by antonfirsov** [comment]
@juliusfriedman users have the knobs now to control platform specific options with `SetRawSocketOption` if they want and really know what they are doing.

However, if we expose a high-level cross-platform option on the public API, those options should be stable, proven, well documented, and  consistent across platforms.

I barely see a way to achieve this for disabling delayed acks. How can we manage it for some undocumented option where we can only guess the behavior based on names in Windows headers and StackOverflow folklore?

**Comment by juliusfriedman** [comment]
I have had them in use in my library for some time and have never had any issues, I would expect the greatest level of compatibility to be able to be achieved on Linux simply due to the open source nature.

The same way we probe for CPU information we should be doing for socket information. Offloading is just as important on the network side in certain cases as it is on the GPU. (Special network links or protocols)

Since you admit that `SetRawSocketOption` is a place for the advanced users what exactly was the purpose of this API proposal, review and approval?

Again you also say a lot of users are asking for it so why not given them a good API around TCP and that will satisfy them especially with easy to use methods which you can control and achieve the same thing as your `delayed ack` which is system wide AFIK and not Socket level in the first place.

**Comment by antonfirsov** [comment]
> Since you admit that `SetRawSocketOption` is a place for the advanced users what exactly was the purpose of this API proposal, review and approval?

This request came before `SetRawSocketOption` existed, the first approval was a mistake (result of incorrect assumptions), the `api approved` label has been removed since then.

> the same thing as your delayed ack which is system wide AFAIK and not Socket level in the first place.

This is not true, delayed ack can be controlled on Socket level in both Linux and Windows (but with very different switches). The goal of this issue is to try to provide a unified API for that specific switch. If you think it's valuable to expose other knobs, I suggest to open a separate issue so the proposal can be discussed and assessed separately.

**Comment by juliusfriedman** [comment]
> > Since you admit that `SetRawSocketOption` is a place for the advanced users what exactly was the purpose of this API proposal, review and approval?
> 
> This request came before `SetRawSocketOption` existed, the first approval was a mistake (result of incorrect assumptions), the `api approved` label has been removed since then.
> 
> > the same thing as your delayed ack which is system wide AFAIK and not Socket level in the first place.
> 
> This is not true, delayed ack can be controlled on Socket level in both Linux and Windows (but with very different switches). The goal of this issue is to try to provide a unified API for that specific switch. If you think it's valuable to expose other knobs, I suggest to open a separate issue so the proposal can be discussed and assessed separately.

That myst be sum switchboard...

**Comment by karelz** [comment]
Moving to Future based on above analysis by @antonfirsov: https://github.com/dotnet/runtime/issues/798#issuecomment-654475815

**Comment by ericsampson** [comment]
This would be great to see in some form in .NET 6

**Comment by antonfirsov** [comment]
It's possible to configure this on .NET 5 on Linux and Windows:

Linux:
```C#
socket.SetRawSocketOption((int) SocketOptionLevel.Tcp, 12, BitConverter.GetBytes(1));
```
Windows: 
```C#
socket.IOControl(SIO_TCP_SET_ACK_FREQUENCY, BitConverter.GetBytes(1), Dummy);
```

I can't see a good cross-platform abstraction for the option though. @ericsampson if you have a real-life use-case for this, is there any problem with the platform-specific solutions I'm suggesting?

**Comment by wfurt** [comment]
triage: workaround exist through platform specific. Since there is no consistent cross-platform it seems like bad fit for general Socket API. We can revisit this in the future.
